L74-java-channel-creds.md

```diff
@@ -0,0 +1,485 @@
+channel/server construction.
+
+This is to allow:
+
+*   implementing new transport credentials without creating a new
+    channel/server type, which allows users of the API to separate concerns
+*   providing stable APIs for advanced credential configuration like mutual
+    TLS, without polluting the general-purpose ManagedChannelBuilder API with
+    many overloads that don't apply to all transports
+*   composition of credentials, like necessary for xDS where a fallback
+    credential needs to be provided by the user
+*   aligning with C core and Go implementations
+
+_The design focuses on channel credentials, since they are the most complex and
+most discussed._ Server credentials will mostly just mirror the channel
+credentials; they suffer from most of the same problems but at a smaller
+scale.
+
+## Background
+
+Prior to gRPC v0.12.0, C core had a "unified" `Credential` type that was
+simultaneously a call and channel credential. This was the design proposed for
+all languages. However, the Java implementation could not determine how such an
+API could be implemented for Java with its multiple transports. The designer of
+the API had too little bandwidth (it was a _very_ busy time), so Java (and Go)
+developed its own API. Java used interceptors (which no other language had) and
+credential-specific configuration on the ManagedChannelBuilder.
+
+In gRPC v0.12.0, C core swapped to a “split” CallCredential and
+ChannelCredential. ChannelCredentials could contain CallCredentials. At this
+point it was not a priority for Java to investigate the new API; Java was happy
+they didn’t use the previous API, such that they didn’t have an API migration.
+
+gRPC Java 0.15.0 added CallCredentials to replace the interceptor-based call
+credentials. At this point Java had looked into the full “split” credential
+design, but decided it was too hard to implement at that moment. Java worked to
+stabilize the `ManagedChannelBuilder.usePlaintext()` and
+`useTransportSecurity()` APIs.
+
+Go has split call credentials (`credentials.PerRPCCredentials`) and channel
+credentials (`credentials.TransportCredentials`), but with a separate type to
+combine them (`credentials.Bundle`). The split credentials significantly
+predate `Bundle`.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+`ChannelCredentials` will be introduced, which can be provided when
+constructing a `ManagedChannelBuilder':
+
+```java
+// instead of:
+channel = ManagedChannelBuilder.forTarget("example.com")
+    .build();
+channel = ManagedChannelBuilder.forTarget("example.com").usePlaintext()
+    .build();
+channel = ManagedChannelBuilder.forAddress("example.com", 443)
+    .build();
+// the user would now:
+channel = Grpc.newChannelBuilder("example.com", new TlsChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilder("example.com", new InsecureChannelCredentials())
+    .build();
+channel = Grpc.newChannelBuilderForAddress("example.com", 443, new TlsChannelCredentials())
```

Comment:
I wonder if we can do away with the `ForAddress` variant in the interest of API simplification (unless there is a benefit I am not seeing).

If the caller wants to specify a port they can always use:
```
newChannelBuilder("example.com:443", new TlsChannelCredentials())
```

Reply:
We have it to deal with IPv6 addresses. IPv6 addresses have to be enclosed in `[]` since they can contain colons. `newChannelBuilderForAddress("::1", 443, creds)` → `newChannelBuilder("[::1]:443", creds)`. `forAddress(host, port)` has always been a convenience, and I'd rather not take it away since I want this API to successfully supplant the existing one.

Reply:
ok